src/Cake.Core/TeardownContext.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using Cake.Core.Diagnostics;
+using Cake.Core.IO;
+using Cake.Core.Tooling;
+
+namespace Cake.Core
+{
+    /// <summary>
+    /// Acts as a context providing info about the overall build following its completion
+    /// </summary>
+    public sealed class TeardownContext : CakeContextAdapter, ITeardownContext
+    {
+        private readonly bool _successful;
+        private readonly Exception _thrownException;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TeardownContext"/> class.
+        /// </summary>
+        /// <param name="cakeContext">The Cake Context</param>
+        /// <param name="successful">If set to <c>true</c>, the build was not successful.</param>
+        /// <param name="throwException">The exception that was thrown by the target.</param>
+        public TeardownContext(ICakeContext cakeContext, bool successful, Exception throwException)
+            : base(cakeContext)
+        {
+            _successful = successful;
```

Comment:
Would a successful build be the same as a build with no exceptions? In that case, maybe we can just check `thrownException` for null instead of providing success. Or do we want this to be explicit? Either one is fine by me, just asking the question loud :smile: